File: Back/La Serranita entradas/src/main/java/org/example/laserranitaentradas/service/CajaService.java
package org.example.laserranitaentradas.service;

import org.example.laserranitaentradas.model.dto.AjusteCajaRequestDTO;
import org.example.laserranitaentradas.model.dto.CajaAbiertaDTO;
import org.example.laserranitaentradas.model.dto.CierrePosnetRequestDTO;
import org.example.laserranitaentradas.model.dto.CajaResponseDTO;
import org.example.laserranitaentradas.model.dto.ConteoDenominacionDTO;
import org.example.laserranitaentradas.model.dto.CajaDetalleAbiertaDTO;
import org.example.laserranitaentradas.model.dto.CajasCerradasResponseDTO;
import org.example.laserranitaentradas.model.entity.Caja;
import org.example.laserranitaentradas.model.entity.TipoMovimientoCaja;
import org.example.laserranitaentradas.model.entity.TipoMovimientoEntradas;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

public interface CajaService {

    /** La caja abierta del usuario, o null si no tiene ninguna en curso. */
    CajaResponseDTO getActual(Long usuarioId);

    /** Devuelve la caja abierta de ese usuario o lanza si no tiene ninguna: la usan las ventas para bloquearse sin caja. */
    Caja getAbiertaOrThrow(Long usuarioId);

    CajaResponseDTO abrir(Long usuarioId, BigDecimal montoInicial, Integer entradasFisicasInicial);

    /**
     * idempotencyKey/fechaOriginal vienen sólo del POS con cola offline: la clave evita duplicar
     * si se reintenta un movimiento cuya respuesta se perdió en un corte, y la fecha registra
     * cuándo pasó de verdad en vez de cuándo se sincronizó. Ambos null desde el uso normal.
     */
    CajaResponseDTO registrarRetiro(Long usuarioId, BigDecimal monto, String motivo, TipoMovimientoCaja tipo,
                                     String idempotencyKey, LocalDateTime fechaOriginal);

    /** Igual que registrarRetiro, pero para que un ADMIN lo cargue en la caja de OTRO usuario (ADMIN-only, gateado en SecurityConfig). Sin cola offline: el admin siempre opera con conexión. */
    CajaResponseDTO registrarRetiroComoAdmin(Long cajaId, BigDecimal monto, String motivo, TipoMovimientoCaja tipo);

    /**
     * Suma o resta entradas físicas al talonario de la caja abierta: INGRESO cuando el boletero
     * se quedó sin y le traen más, RETIRO para sacarle entradas y dárselas a otro boletero (motivo
     * opcional). Ver registrarRetiro por idempotencyKey/fechaOriginal.
     */
    CajaResponseDTO registrarIngresoEntradas(Long usuarioId, Integer cantidad, String motivo, TipoMovimientoEntradas tipo,
                                              String idempotencyKey, LocalDateTime fechaOriginal);

    /** Igual que registrarIngresoEntradas, pero para que un ADMIN lo cargue en la caja de OTRO usuario por id (ADMIN-only, gateado en SecurityConfig). Sin cola offline. */
    CajaResponseDTO registrarIngresoEntradasComoAdmin(Long cajaId, Integer cantidad, String motivo, TipoMovimientoEntradas tipo);

    CajaResponseDTO cerrar(Long usuarioId, List<ConteoDenominacionDTO> conteoEfectivo,
                            List<CierrePosnetRequestDTO> cierresPosnet, Integer entradasFisicasCortadas,
                            BigDecimal cambioContado, BigDecimal dolaresContado);

    /** Igual que cerrar, pero para que un ADMIN cierre la caja de OTRO usuario por id (ADMIN-only, gateado en SecurityConfig). */
    CajaResponseDTO cerrarComoAdmin(Long cajaId, List<ConteoDenominacionDTO> conteoEfectivo,
                                     List<CierrePosnetRequestDTO> cierresPosnet, Integer entradasFisicasCortadas,
                                     BigDecimal cambioContado, BigDecimal dolaresContado);

    /**
     * Corrige los datos de un cierre ya hecho (ej. un billete mal contado). Sólo si ya está
     * cerrada — no reemplaza a `cerrar`. ADMIN-only (gateado en SecurityConfig): no valida
     * dueño acá, cualquier caja cerrada se puede corregir.
     */
    CajaResponseDTO corregirCierre(Long cajaId, List<ConteoDenominacionDTO> conteoEfectivo,
                                    List<CierrePosnetRequestDTO> cierresPosnet, Integer entradasFisicasCortadas,
                                    BigDecimal cambioContado, BigDecimal dolaresContado);

    /**
     * Reabre momentáneamente una caja ya cerrada (fechaCierre → null) para poder aplicarle una
     * operación que se había rechazado porque esa caja ya no estaba abierta — retiro/aporte,
     * ingreso de entradas, o una venta. No aplica ningún movimiento por sí sola: eso lo hace
     * quien llama, usando los métodos normales de siempre (registrarRetiro,
     * registrarIngresoEntradas, CompraService.registrarVentaPos) contra ESTA caja ya reabierta
     * — así cada operación se revalida en vivo con las reglas reales (cupo, promo vigente, etc.)
     * en vez de forzarse a ciegas. Lanza si la caja no existe o si sigue abierta (nada que
     * reabrir). ADMIN-only (gateado en SecurityConfig, vía RechazoOperacionController), que es
     * quien orquesta reabrir → reaplicar (una o varias operaciones de la misma caja) →
     * recerrarConElUltimoConteo.
     */
    Caja reabrir(Long cajaId);

    /**
     * Vuelve a cerrar una caja que se reabrió con reabrir(), reutilizando el mismo conteo
     * (denominaciones, posnet, entradas cortadas, cambio, dólares) que ya tenía guardado de
     * antes — no hace falta pedirle a nadie que cuente billetes de nuevo por operaciones que ni
     * siquiera vio. Lanza si la caja no existe o si ya está cerrada.
     */
    CajaResponseDTO recerrarConElUltimoConteo(Long cajaId);

    /** Detalle completo de cualquier caja (para que el admin la revise sin importar quién la abrió). */
    CajaResponseDTO getDetalle(Long cajaId);

    /** Todas las cajas abiertas ahora mismo, sin importar de qué boletero — para el dashboard del admin. */
    List<CajaAbiertaDTO> getCajasAbiertas();

    /**
     * Ventas, retiros/aportes e ingresos de entradas de una caja, en orden cronológico, más un
     * resumen de lo vendido hasta el momento (por forma de pago y por tipo de entrada) — a
     * diferencia de getDetalle(), funciona con la caja todavía ABIERTA (no espera al cierre):
     * lo usa el admin para revisar y corregir una venta mal cargada mientras el boletero sigue
     * trabajando. ADMIN-only (gateado en SecurityConfig): a un boletero no se le puede mostrar
     * esto de su propia caja en curso, se prestaría a ajustar el conteo para que cierre justo.
     */
    CajaDetalleAbiertaDTO getOperacionesCaja(Long cajaId);

    /**
     * Cajas cerradas dentro del rango, paginadas y opcionalmente filtradas por boletero — para el
     * listado de "Cajas cerradas" en la pantalla de Cajas. A diferencia del reporte agregado
     * (ReporteService.getResumen, que trae TODAS las cajas del rango de una para calcular varios
     * KPIs a la vez), esto pagina en la base: soporta boleteros con meses de turnos sin traerlos
     * todos a memoria. Los totales de retiros/faltantes/sobrantes son de TODO lo que matchea el
     * filtro, no sólo la página (ver CajasCerradasResponseDTO). ordenarPor admite cualquier campo
     * de CajaResumenReporteDTO excepto "totalRetiros" (se computa con un JOIN + SUM, no es una
     * columna propia de Caja para ordenar en la base sin recorrer todo el rango).
     */
    CajasCerradasResponseDTO getCajasCerradas(LocalDate desde, LocalDate hasta, String usuarioNombre,
                                               String ordenarPor, String direccion, int page, int size);

    /**
     * Registra uno o más traspasos manuales de monto entre formas de pago sobre una caja YA
     * CERRADA (el admin corrigiendo la repartición cuando la cajera cobró de una forma y tocó
     * otra). No toca las compras: cada traspaso queda como un AjusteCaja aparte. Recalcula y
     * persiste montoEsperado/diferencia para que el listado de cajas cerradas y el ranking
     * queden consistentes. ADMIN-only (gateado en SecurityConfig).
     */
    CajaResponseDTO registrarAjustes(Long cajaId, List<AjusteCajaRequestDTO> ajustes);

    /** Deshace un ajuste manual: borra la fila y recalcula el cierre. ADMIN-only. */
    CajaResponseDTO eliminarAjuste(Long cajaId, Long ajusteId);
}

File: Front/src/app/services/caja.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { etiquetaFormaPago } from '../models/forma-pago';
import { Pagina } from './boleteria.service';

export type TipoMovimientoCaja = 'RETIRO' | 'APORTE';

export type TipoMovimientoEntradas = 'INGRESO' | 'RETIRO';

export interface RetiroCaja {
  id: number;
  monto: number;
  motivo: string;
  tipo: TipoMovimientoCaja;
  fecha: string;
}

export interface ConteoDenominacion {
  denominacion: number;
  cantidad: number;
}

export type FormaPagoPosnet = 'TARJETA' | 'MERCADO_PAGO_QR';

export interface CierrePosnet {
  id: number;
  /** Null = cargado combinado (Tarjeta+QR juntos, sin distinguir). */
  formaPago: FormaPagoPosnet | null;
  monto: number;
  nota: string | null;
  fecha: string;
}

export interface IngresoEntradas {
  id: number;
  cantidad: number;
  motivo: string | null;
  tipo: TipoMovimientoEntradas;
  fecha: string;
}

export interface OperacionCaja {
  tipo: 'VENTA' | 'RETIRO' | 'APORTE' | 'INGRESO_ENTRADAS' | 'RETIRO_ENTRADAS';
  fecha: string;
  /** Null en los ingresos de entradas físicas: no mueven plata. */
  monto: number | null;
  formaPago: string | null;
  detalle: string;
  /** Sólo en ventas: id de la Compra, para poder cancelarla o editarla. Null en retiros/ingresos. */
  compraId: number | null;
}

/** Nombre corto para mostrar en el detalle de caja: la forma de pago cruda del backend es un enum, no algo para mostrar tal cual. */
export function etiquetaTipoOperacion(op: OperacionCaja): string {
  if (op.tipo === 'VENTA') return etiquetaFormaPago(op.formaPago);
  if (op.tipo === 'INGRESO_ENTRADAS') return 'Ingreso entradas';
  if (op.tipo === 'RETIRO_ENTRADAS') return 'Retiro entradas';
  if (op.tipo === 'APORTE') return 'Aporte';
  return 'Retiro';
}

/**
 * Mientras la caja está ABIERTA, el backend manda todos los totales "esperados" (efectivo,
 * tarjeta, QR, entradas físicas) y el detalle de operaciones en null a propósito: si el
 * boletero pudiera verlos antes de cerrar, el cierre dejaría de ser un control real. Recién
 * llegan completos en la respuesta del cierre.
 */
export interface Caja {
  id: number;
  estado: 'ABIERTA' | 'CERRADA';
  fechaApertura: string;
  montoInicial: number;
  fechaCierre: string | null;
  totalVentasEfectivo: number | null;
  totalRetiros: number;
  efectivoEsperado: number | null;
  montoContado: number | null;
  diferencia: number | null;
  /** Total en billetes chicos cargado de una vez; ya está incluido en montoContado. Null hasta el cierre. */
  cambioContado: number | null;
  retiros: RetiroCaja[];

  conteoEfectivo: ConteoDenominacion[];

  totalVentasTarjeta: number | null;
  totalVentasQr: number | null;
  totalCerradoTarjeta: number | null;
  totalCerradoQr: number | null;
  diferenciaTarjeta: number | null;
  diferenciaQr: number | null;
  /** Si el cierre se cargó combinado (Tarjeta+QR juntos), estos tres reemplazan a los de arriba; si no, quedan null. */
  totalVentasPosnet: number | null;
  totalCerradoPosnet: number | null;
  diferenciaPosnet: number | null;
  cierresPosnet: CierrePosnet[];

  entradasFisicasInicial: number | null;
  entradasFisicasCortadas: number | null;
  entradasFisicasEsperadas: number | null;
  diferenciaEntradas: number | null;
  totalIngresosEntradas: number;
  ingresosEntradas: IngresoEntradas[];

  operaciones: OperacionCaja[] | null;

  /** Unidades de entrada vendidas (no extras ni artículos), sin importar la forma de pago. Null hasta el cierre. */
  totalEntradasVendidas: number | null;
  entradasVendidasPorTipo: EntradasPorTipo[] | null;

  /**
   * Si hubo o no alguna venta pagada en dólares (sigue siendo un cobro en efectivo, sólo
   * cambia la moneda física). A diferencia de los totales, esto se expone aunque la caja
   * siga ABIERTA (es sólo un booleano, no un monto).
   */
  huboVentaDolares: boolean;
  /** Dólares esperados según lo que entró de cada venta en dólares. Null hasta el cierre. */
  dolaresEsperado: number | null;
  /** Dólares que el boletero contó al cerrar. Null si esta caja no tuvo ventas en dólares. */
  dolaresContado: number | null;
  diferenciaDolares: number | null;
}

export interface EntradasPorTipo {
  nombreTipo: string;
  cantidad: number;
}

export interface CierrePosnetInput {
  /** Null = combinado (Tarjeta+QR juntos, sin distinguir). */
  formaPago: FormaPagoPosnet | null;
  monto: number;
  nota?: string | null;
}

/** Una caja abierta ahora mismo, para el dashboard de hoy del admin. */
export interface CajaAbierta {
  id: number;
  usuarioNombre: string;
  fechaApertura: string;
  montoInicial: number;
  /** Vendido hasta el momento (efectivo + tarjeta + QR), en vivo. */
  totalVendido: number;
  totalEntradasVendidas: number;
}

/** Detalle de una caja todavía abierta (ADMIN) — ver obtenerOperaciones. */
export interface CajaDetalleAbierta {
  operaciones: OperacionCaja[];
  totalVentasEfectivo: number;
  totalVentasTarjeta: number;
  totalVentasQr: number;
  totalEntradasVendidas: number;
  entradasVendidasPorTipo: EntradasPorTipo[];
  huboVentaDolares: boolean;
  /** Inicial + ingresos − retiros − ya cortadas vendiendo: cuántas le quedan al boletero en el talonario. Null si esta caja no tiene un inicial cargado. */
  entradasFisicasRestantes: number | null;
}

/** Una caja ya cerrada, para el listado paginado de "Cajas cerradas" (ver obtenerCajasCerradas). */
export interface CajaCerrada {
  id: number;
  usuarioNombre: string;
  fechaApertura: string;
  fechaCierre: string;
  montoInicial: number;
  totalRetiros: number;
  montoEsperado: number;
  montoContado: number;
  diferencia: number;
}

/** Página de "Cajas cerradas": mismas propiedades que Pagina<T>, más los totales de retiros/
 * faltantes/sobrantes de TODO lo que matchea el filtro (no sólo la página actual). */
export interface CajasCerradasResponse extends Pagina<CajaCerrada> {
  totalRetiros: number;
  totalFaltantes: number;
  totalSobrantes: number;
}

@Injectable({
  providedIn: 'root',
})
export class CajaService {
  private http = inject(HttpClient);
  private cajaUrl = `${environment.apiBase}/interno/caja`;

  /** La caja abierta del boletero autenticado, o null si no tiene ninguna en curso. */
  getActual(): Observable<Caja | null> {
    return this.http.get<Caja | null>(`${this.cajaUrl}/actual`);
  }

  abrir(montoInicial: number, entradasFisicasInicial: number): Observable<Caja> {
    return this.http.post<Caja>(`${this.cajaUrl}/abrir`, { montoInicial, entradasFisicasInicial });
  }

  /** idempotencyKey/fechaOriginal/esReintentoEncolado sólo los manda la cola offline del POS
   * (ver OperacionesPendientesService). cajaId es la caja propia del boletero en ese momento:
   * no se usa para resolver dónde aplicar el movimiento (eso lo sigue derivando el backend del
   * usuario autenticado), sólo queda guardado en el rechazo si esto se rechaza, para saber
   * exactamente qué caja reabrir y reintentar. */
  registrarRetiro(
    monto: number,
    motivo: string,
    tipo: TipoMovimientoCaja,
    idempotencyKey?: string,
    fechaOriginal?: string,
    esReintentoEncolado?: boolean,
    cajaId?: number
  ): Observable<Caja> {
    return this.http.post<Caja>(`${this.cajaUrl}/retiros`, { monto, motivo, tipo, idempotencyKey, fechaOriginal, esReintentoEncolado, cajaId });
  }

  /** Igual que registrarRetiro, pero un ADMIN cargándolo en la caja de OTRO usuario (ej. mientras la está cerrando). */
  registrarRetiroComoAdmin(cajaId: number, monto: number, motivo: string, tipo: TipoMovimientoCaja): Observable<Caja> {
    return this.http.post<Caja>(`${this.cajaUrl}/${cajaId}/retiros`, { monto, motivo, tipo });
  }

  registrarIngresoEntradas(
    cantidad: number,
    tipo: TipoMovimientoEntradas,
    motivo?: string,
    idempotencyKey?: string,
    fechaOriginal?: string,
    esReintentoEncolado?: boolean,
    cajaId?: number
  ): Observable<Caja> {
    return this.http.post<Caja>(`${this.cajaUrl}/ingresos-entradas`, { cantidad, tipo, motivo, idempotencyKey, fechaOriginal, esReintentoEncolado, cajaId });
  }

  /** Igual que registrarIngresoEntradas, pero un ADMIN cargándolo en la caja de OTRO usuario por id. */
  registrarIngresoEntradasComoAdmin(cajaId: number, cantidad: number, motivo: string | undefined, tipo: TipoMovimientoEntradas): Observable<Caja> {
    return this.http.post<Caja>(`${this.cajaUrl}/${cajaId}/ingresos-entradas`, { cantidad, motivo, tipo });
  }

  /** Cerrar caja es ADMIN-only (ya no self-service): cierra la caja de cualquier usuario por id. */
  cerrarComoAdmin(
    cajaId: number,
    conteoEfectivo: ConteoDenominacion[],
    cierresPosnet: CierrePosnetInput[],
    entradasFisicasCortadas: number,
    cambioContado: number | null,
    dolaresContado: number | null
  ): Observable<Caja> {
    return this.http.post<Caja>(`${this.cajaUrl}/${cajaId}/cerrar`, {
      conteoEfectivo,
      cierresPosnet,
      entradasFisicasCortadas,
      cambioContado,
      dolaresContado,
    });
  }

  /** Detalle completo de cualquier caja (ADMIN), sin importar quién la abrió. */
  obtenerDetalle(cajaId: number): Observable<Caja> {
    return this.http.get<Caja>(`${this.cajaUrl}/${cajaId}/detalle`);
  }

  /** Todas las cajas abiertas ahora mismo (ADMIN), para el dashboard de hoy. */
  obtenerCajasAbiertas(): Observable<CajaAbierta[]> {
    return this.http.get<CajaAbierta[]>(`${this.cajaUrl}/abiertas`);
  }

  /** Ventas, retiros/aportes e ingresos de entradas de una caja, más un resumen de lo vendido
   * (ADMIN) — a diferencia de obtenerDetalle, funciona con la caja todavía abierta. */
  obtenerOperaciones(cajaId: number): Observable<CajaDetalleAbierta> {
    return this.http.get<CajaDetalleAbierta>(`${this.cajaUrl}/${cajaId}/operaciones`);
  }

  /** Cajas cerradas dentro del rango, paginadas y opcionalmente filtradas por boletero (ADMIN):
   * soporta boleteros con meses de turnos sin traerlos todos de una. ordenarPor admite cualquier
   * campo de CajaCerrada salvo "totalRetiros" (no es una columna propia de Caja). */
  obtenerCajasCerradas(
    desde: string,
    hasta: string,
    usuarioNombre: string | null,
    ordenarPor: string,
    direccion: 'ASC' | 'DESC',
    page: number,
    size: number
  ): Observable<CajasCerradasResponse> {
    let params = new HttpParams()
      .set('desde', desde)
      .set('hasta', hasta)
      .set('ordenarPor', ordenarPor)
      .set('direccion', direccion)
      .set('page', page)
      .set('size', size);
    if (usuarioNombre) params = params.set('usuarioNombre', usuarioNombre);
    return this.http.get<CajasCerradasResponse>(`${this.cajaUrl}/cerradas`, { params });
  }

  /** Corrige un cierre ya hecho (ej. un billete mal contado). ADMIN-only, cualquier caja cerrada. */
  corregirCierre(
    cajaId: number,
    conteoEfectivo: ConteoDenominacion[],
    cierresPosnet: CierrePosnetInput[],
    entradasFisicasCortadas: number,
    cambioContado: number | null,
    dolaresContado: number | null
  ): Observable<Caja> {
    return this.http.put<Caja>(`${this.cajaUrl}/${cajaId}/cierre`, {
      conteoEfectivo,
      cierresPosnet,
      entradasFisicasCortadas,
      cambioContado,
      dolaresContado,
    });
  }
}
